### DISCUSSION TITLE
WinUSB instead of VBOXUsb drivers

### DISCUSSION THREAD

**Discussion by sidd-kishan** [dorssel/usbipd-win#374] [Ideas]
Is it possible to replace the usage of WinUsb instead of VBOXUsb drivers as a way to provide compatibility with the ARM64 platform?

**Comment by dorssel** [comment]
Not really. I looked into this an came to the following conclusions:
1) WinUSB requires specific support in the device's USB descriptors to work as intended, although (with a hack) you can make it load for any device.
2) WinUSB does not allow for devices with multiple configurations (simply not supported), so some devices won't work at all.
3) Devices that are themselves not WinUSB compatible (see point 1), need a modified INF file to register an interface such that the device is actually usable. This invalidates the CAT signature, which in turn requires resigning the driver.

And that last point is exactly the problem: driver signatures. The *only* reason I chose VBoxUSB is that the driver is properly signed. If I could sign my own (i.e., if I would have an EV-cert and Microsoft would accept individuals to register in the signing portal), then I would have written a specific driver myself. I was able to make the one from VBox work (although it is kind of poorly written, judging from the code) *and* it is already signed. WinUSB is closed source, and the documented interface is not enough for all types of devices *and* would require re-signing for many devices.

So, VBoxUSB is the only way (that I see), for now. I would write a better driver myself, if only I could get it signed...

**Comment by Tremek** [comment]
Hi, any updates on possibilities for Win 11 on ARM (Windows Dev Kit 2023 specifically) support? I’m trying to pass a USB mic through to WSL2 Debian and obviously stymied. You mention getting a driver signed is the primary issue: is it just due to cost of getting an EV cert?

**Reply by dorssel** [reply]
There are several reasons:

- Writing a driver takes a considerable amount of time. I can do it, but it is a little more effort then I would like for a hobby project.
- Getting a driver *stable* takes even more time. E.g., even though the VirtualBox drivers are not of very high quality, they have been around for two decades or so. It takes a while to reach the same amount of stability.
- EV costs, yes, that too
- Microsoft does not allow individuals (only organizations) to partner in their hardware program. I would have to "cheat" to bypass that.

So, my current plans do not include replacing the driver.

**Reply by sidd-kishan** [reply]
Can we explore the possiblity of sponsor to support the feature of windows on ARM64 project through crowd funding. Please tell.

**Reply by dorssel** [reply]
Not for me. But if anyone is able to provide a signed arm64 compile of the VirtualBox drivers, I'll be happy to integrate them and make an arm64 version available.

**Comment by Tremek** [comment]
One more question; is it possible that you could provide the VS .sln etc. And allow those of us silly enough to want to attempt this on aarch64 hardware to self-sign and compile for ourselves in the meantime?

**Reply by dorssel** [reply]
???
You're on GitHub. All the source is already there.

**Comment by dorssel** [comment]
ARM64 support is part of version 5.0.0, which has just been released.
